Adds prototype registry storage

xfbs opened the pull request:
This PR adds a prototype registry storage implementation. This uses a trait to be generic against the storage provider, and a trait object to minimize code size and compilation time.

The goal of the storage module is to put and get the packages from a storage system such as S3 or artifactory.

The next step here is to add the metadata/database trait. What makes this more complex is that we might want to support multiple schemes, such as using a database or using artifactory with some "hacks" to store the things we need within artifact labels.

As this is something that we need to explore, I am creating this pull request so that we can already get the storage side of things up and running.

xfbs commented:
Current status:

- I have moved all tests into the `registry/tests` folder, as @mara-schulke had requested.
- I have added CI infrastructure to tests the registry.
- Cargo deny is broken, I dunno why (can we just turn it off or configure it?)
- Coverage for registry is broken, llvm-cov does not pass because it counts `buffrs` code in it's coverage calculation erroneously. I don't know for sure how to fix this, maybe we can just turn off coverage calculation for now :cry:

Feel free to review this, also feel free to push to this branch to fix things if you like!

Cheers,
@xfbs

mara-schulke commented on registry/Cargo.toml:
```diff
@@ -16,6 +17,36 @@ tracing-subscriber = "0.3.17"
 sqlx = { version = "0.7.2", features = ["runtime-tokio", "postgres", "macros", "migrate"] }
 eyre = "0.6.8"
 url = "2.4.1"
+axum_tonic = "0.1.0"
```
Super nit picky / housekeeping related, but could we eventually sort the deps? I can also do it after the merge

xfbs replied:
Of course! By the way, if this is something we care about, we can even enforce it in CI with [cargo sort](https://crates.io/crates/cargo-sort/1.0.5).

mara-schulke commented on registry/src/bin/database.rs:
```diff
@@ -0,0 +1,4 @@
+//use buffrs_registry::*;
+
+#[tokio::main(flavor = "current_thread")]
+async fn main() {}
```
Why do we have a bin "database"?

xfbs replied:
Honestly, I don't know what I was doing there :D I probably used that file for some quick tests and forgot to delete it.

mara-schulke commented on registry/src/storage/tests.rs:
```diff
@@ -0,0 +1,138 @@
+// Copyright 2023 Helsing GmbH
```
Can this file live in "registry/tests/storage.rs"?

mara-schulke replied:
(Following the rule of IO related tests being declared outside of the source code)

xfbs replied:
For sure, this file is actually unused (I just forgot to delete it).

mara-schulke commented on registry/src/types.rs:
```diff
@@ -0,0 +1,124 @@
+    }
+}
+
+prop_compose! {
```
This doesn't seem like it should live here

xfbs replied:
Removed it!

mara-schulke commented on registry/src/types.rs:
```diff
@@ -0,0 +1,124 @@
+    // FIXME(xfbs): use PackageName here
+    /// Package name
+    #[strategy(package_name())]
+    pub package: Arc<str>,
```
This is a PackageNamr in the buffrs lib iirc

xfbs replied:
Fixed that — now using `PackageName` and `semver::Version` here.

mara-schulke commented on registry/src/types.rs:
```diff
@@ -0,0 +1,124 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use proptest::prelude::*;
```
I dislike that proptest is mixed into the source code here - any other way around?

xfbs replied:
Currently, what I am trying to do is derive the `Arbitrary` trait on things like `PackageVersion`. This allows me to write tests with proptest like so:

```rust
#[proptest]
async fn some_test(arbitrary_version: PackageVersion) {
    // ...
}
```

This does mean that we need to have some amount of proptest-related code in here (to derive that). Unfortunately, we cannot put it elsewhere because of the orphan rule (we cannot derive `Arbitrary` in the `tests/` for example). 

There is another way, which is using manual proptest strategies that we could employ which mean that we do not need to derive `Arbitrary`. It would look like this:

```
prop_compose! {
    fn package_version() -> PackageVersion {
        // create package version
    }
}

#[proptest]
async fn some_test(arbitrary_version in package_version()) {
    // ...
}
```

Now we could put that `prop_compose` into the tests folder.

mara-schulke replied:
i would put the prop compone in the tests folder!
for the derive – can we feature flag it with the `test` feature, so its clear that this is for testing purposes only?

xfbs replied:
We have moved all of the tests into `tests/` folder, which means that they are integration tests. Integration tests are not built with the `test` feature (they compile the crate exactly as it would be compiled by a user). So, unfortunately we do not have any flag we can check :/

I think the only option is to simply not derive `Arbitrary`, but implement the strategy manually.

xfbs replied:
Fixed this by not deriving Arbitrary.

mara-schulke reviewed (commented):
In general I think this is good. I still disagree the existence of the cache implementation (even if it works / is simple) because this is just code to maintain that we won't need for a very long time. 

Other than that I left a few comments on the structure of the project (i.e where the tests should live and separation of source and test code)

What do you think about the caching topic @xfbs?

xfbs commented:
> What do you think about the caching topic @xfbs?

I don't really have any strong opinions, I think it is neat to have but with the discussions we have had about artifactory, it sounds like we might not be able to use a cache for now anways, so I have removed it. At least the code is written in a way that should make it easy for us to add a cache should we want or need to at some point. We will definitely need it if we ever deploy a public instance of it (which is something that I would be interested in, at some point).

Besides that: I have just fixed the remaining CI issues (replaced `tempdir` with `tempfile`). So at least on the CI front, it is looking good at the moment :)

mara-schulke commented on registry/Cargo.toml:
```diff
@@ -4,18 +4,47 @@ version = "0.1.0"
+[features]
+default = ["storage-s3"]
+storage-s3 = ["dep:aws-config", "dep:aws-sdk-s3"]
+test-database = []
```
Why do we have this? it doesnt seem good to publish this feature as it seems like a local dev artifact

xfbs replied:
You are right — will remove this.

mara-schulke commented on registry/src/storage.rs:
```diff
@@ -0,0 +1,109 @@
+mod s3;
+
+#[cfg(feature = "storage-cache")]
+pub use cache::{Cache, CacheConfig};
```
Still an artifact from the cache impl?

xfbs replied:
Yes 😬 I will remove this.

mara-schulke commented:
Left a final set of things that looked odd to me – you can resolve them if you want – and merge!

xfbs commented:
I did not get around to formatting the SQL stuff — I tried using [sleek](https://crates.io/crates/sleek), but it does not fully solve our problem because it only does the formatting but does not get the style consistent, so I'll address this in a future PR!
